### ISSUE TITLE
PP and non-PP produce differently scaled gradients

### ISSUE THREAD

**Issue by wwwjn** [pytorch/torchtitan#2972]
## Summary

Pipeline Parallel (PP) and non-PP training produce differently scaled gradients for the same model and data. In non-PP, the loss is divided by `global_valid_tokens` before backward, so gradients are normalized. In PP, the schedule backwards the raw sum loss and the `/ global_valid_tokens` division happens after backward (for metrics only). This means PP gradients are `global_valid_tokens` times larger than non-PP gradients.

## Root Cause

**Non-PP path** (`trainer.py`):
```python
loss_sum = self.loss_fn(pred, labels)       # sum reduction
loss = loss_sum / global_valid_tokens       # scale BEFORE backward
loss.backward()                             # gradients are normalized
```

**PP path** (`trainer.py` + PyTorch `schedules.py`):
```python
# Inside pp_schedule.step(), on the last stage:
#   schedules.py:285 — loss = loss_fn(output, target)        # sum reduction, unscaled
#   schedules.py:766 — stage.backward_one_chunk(loss=loss)   # gradients are NOT normalized

# After pp_schedule.step(), in trainer.py:
loss = torch.sum(torch.stack(losses)) / global_valid_tokens  # metrics only, backward already done
```

Additionally:
- `scale_grads=False` is set in `build_pipeline_schedule` (`pipeline_parallel.py:234`)
- `gradient_divide_factor=1.0` via `disable_fsdp_gradient_division`

So neither the PP schedule nor FSDP normalizes the gradients.

## Reproduce

```bash
# Non-PP: 4 GPUs, FSDP only
NGPU=4 LOG_RANK=0 ./run_train.sh \
  --training.steps 3 --debug.seed 42 --debug.deterministic \
  --dump_folder /tmp/grad_scale_test/non_pp

# PP: 4 GPUs, PP2 + FSDP2
NGPU=4 LOG_RANK=0,2 ./run_train.sh \
  --training.steps 3 --debug.seed 42 --debug.deterministic \
  --parallelism.pipeline_parallel_degree 2 \
  --dump_folder /tmp/grad_scale_test/pp
```

Compare `grad_norm` between the two runs:

| Step | Non-PP grad_norm | PP grad_norm | Ratio |
|------|-----------------|-------------|-------|
| 1 | 1.5181 | 50,291.5 | ~33,000x |
| 2 | 1.5738 | 53,092.8 | ~33,700x |
| 3 | 1.9868 | 67,011.1 | ~33,700x |

The ratio matches `global_valid_tokens` (`local_batch_size=8 * seq_len=2048 * dp=2 = 32,768`, minus IGNORE_INDEX tokens).

### Versions

torchtitan main

**Comment by tianyu-l** [comment]
Should PP allow setting per-iteration loss, at least per-iteration loss scaling factor? @sanketpurandare @acisseJZhong

**Comment by mirceamironenco** [comment]
An equivalent and possibly less invasive alternative for the PP case would be to scale the gradients by `1 / global_valid_tokens` before gradient clipping (i.e. after all gradient-accumulation microbatches have finished). An optimized (subsequent) version could mirror the PT `clip_grads_with_norm_` implementation (i.e. have a standard and a `torch._foreach_mul_` path).

I’d be happy to contribute this approach if it’s preferred.

**Comment by wwwjn** [comment]
> to scale the gradients by `1 / global_valid_tokens` before gradient clipping

We discussed this possible path while implementing the loss scaling. Mathematically it would work, but it might causing gradient overflow in actual run?

**Comment by mirceamironenco** [comment]
That's probably the more numerically stable approach.

I might be wrong but I believe right now the validator reuses the pp_schedule and loss, expecting raw sums, so if the scale from the previous training step is not set to None, it might be dividing twice. I guess one could set it to None after every optimizer.step() since the mutated state might affect other components.

**Comment by wwwjn** [comment]
> That's probably the more numerically stable approach.
> 
> I might be wrong but I believe right now the validator reuses the pp_schedule and loss, expecting raw sums, so if the scale from the previous training step is not set to None, it might be dividing twice. I guess one could set it to None after every optimizer.step() since the mutated state might affect other components.

Thanks for bring up the validator. The limitation of current fix is `ScaledLoss` is stateful. I think a better solution is to make Pipeline parallel code more flexible and take a scaling factor

**Comment by wwwjn** [comment]
This should be fixed in #2937, which refactored and put the loss refactor in loss function

**Comment by wwwjn** [comment]
Solved together in #2937
